### PR TITLE
Add user-defined cluster metadata

### LINKED ISSUE #33220: Add User-Defined Cluster Metadata
It would be useful to have user-defined, cluster-wide metadata can be set and accessed by users, particularly users who run multiple clusters.  Example use cases include:

1. An organization with multiple clusters, each under a different administrator, could set `"administrator": "joe@example.com"` to quickly identify the owner of each cluster.
2. An organization with different clusters for Testing and Production could set `"environment": "test"` in the Testing cluster and verify in a test runner that this property is correctly set before running tests against that cluster.

We will address this by allowing users to use the [Cluster Settings API](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-update-settings.html) to set arbitrary key/value pairs with a specific key prefix.  For example, if the key prefix is `cluster.metadata`, the example keys above would be `cluster.metadata.administrator` and `cluster.metadata.environment`.

These settings have no semantic meaning to them from Elasticsearch itself, and only serve to store information users wish to associate with their cluster.

### PR DESCRIPTION
**Pull request by AthenaEryma**
Adds a place for users to store cluster-wide data they wish to associate
with the cluster via the Cluster Settings API. This is strictly for
user-defined data, Elasticsearch makes no other other use of these
settings.

Closes #33220

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/cluster/settings/ClusterSettingsIT.java [hunk 1]
```diff
@@ -380,4 +381,23 @@ public void testLoggerLevelUpdate() {
...
+        String value = randomRealisticUnicodeOfCodepointLengthBetween(5, 50);
+
+        // Make sure we have both key and value in case this fails
+        logger.info("Attempting to store [{}]: [{}]", key, value);
+
+        final Settings settings = Settings.builder().put(key, value).build();
+        if (randomBoolean()) {
+            client().admin().cluster().prepareUpdateSettings().setPersistentSettings(settings).execute().actionGet();
+            ClusterStateResponse state = client().admin().cluster().prepareState().execute().actionGet();
+            assertEquals(value, state.getState().getMetaData().persistentSettings().get(key));
```
Comment (s1monw) [inline server/src/test/java/org/elasticsearch/cluster/settings/ClusterSettingsIT.java]: can you also add a tests that updates a setting

File: server/src/test/java/org/elasticsearch/cluster/settings/ClusterSettingsIT.java [hunk 2]
```diff
@@ -380,4 +381,23 @@ public void testLoggerLevelUpdate() {
         }
     }
 
+    public void testUserMetadata() {
+        String key = "cluster.metadata." + randomAlphaOfLengthBetween(5, 20);
+        String value = randomRealisticUnicodeOfCodepointLengthBetween(5, 50);
+
+        // Make sure we have both key and value in case this fails
+        logger.info("Attempting to store [{}]: [{}]", key, value);
```
Comment (s1monw) [inline server/src/test/java/org/elasticsearch/cluster/settings/ClusterSettingsIT.java]: this logging statement is unnecessary IMO.

**Review by s1monw — APPROVED** [review]
Left 2 comments. LGTM otherwise

File: server/src/test/java/org/elasticsearch/cluster/settings/ClusterSettingsIT.java (hunk 1 above)
Comment (gwbrown) [reply server/src/test/java/org/elasticsearch/cluster/settings/ClusterSettingsIT.java]: Will do.

File: docs/reference/modules/cluster/misc.asciidoc [hunk 3]
```diff
@@ -22,6 +22,27 @@ user with access to the <<cluster-update-settings,cluster-update-settings>>
...
+[[user-defined-data]]
+==== User Defined Cluster Metadata
+
+User-defined metadata can be stored and retrieved using the Cluster Settings API.
+This can be used to store arbitrary, infrequently-changing data about the cluster
+without the need to create an index to store it. This data may be stored using
+any key prefixed with `cluster.metadata.`.  For example, to store the email
+address of the administrator of a cluster under the key `cluster.metadata.administrator`,
+issue this request:
+
```
Comment (mayya-sharipova) [inline docs/reference/modules/cluster/misc.asciidoc]: just some questions for the user's perspective: 

> This data may be stored using

may be? or must be?

Also, could be worth mentioning that settings can be both persistent and transient

**Review by mayya-sharipova — APPROVED** [review]

**Review by jasontedor — APPROVED** [review]
This LGTM to me as a v1. We might have to consider adding dedicated APIs for viewing these, or surfacing them in existing APIs (for example, `/`), but let's get this in and see what comes.

File: docs/reference/modules/cluster/misc.asciidoc (hunk 3 above)
Comment (gwbrown) [reply docs/reference/modules/cluster/misc.asciidoc]: I think most (all?) settings can be set either/both `persistent` or `transient` - at least, that's how I read the [Cluster Settings API docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-update-settings.html), so I don't think it's critical to mention that here.

**Pull request merged**